src/ess/spectroscopy/indirect/kf.py

```diff
@@ -93,138 +87,190 @@ def sample_analyzer_vector(
     from scipp.constants import hbar, neutron_mass
 
-    velocity = kf_magnitude * hbar / neutron_mass
-    return secondary_flight_distance / velocity
+    velocity = final_wavenumber * (hbar / neutron_mass)
+    return sc.to_unit(L2 / velocity, 'ms', copy=False)
+
 
+def secondary_spectrometer_coordinate_transformation_graph(
+    analyzer: Analyzer[RunType],
+) -> SecondarySpecCoordTransformGraph[RunType]:
+    """Return a coordinate transformation graph for the secondary spectrometer.
 
-def sample_frame_time(
-    detector_time: DetectorFrameTime, secondary_time: SampleDetectorFlightTime
-) -> SampleFrameTime:
-    """Return the time each neutron likely interacted with the sample"""
-    return detector_time - secondary_time
+    Parameters
+    ----------
+    analyzer:
+        Data group with analyzer parameters.
+
+    Returns
+    -------
+    :
+        Coordinate transformation graph for the secondary spectrometer.
+        The graph captures the relevant parameters of ``analyzer``.
+    """
+    from scippneutron.conversion.beamline import beam_aligned_unit_vectors
+
+    return SecondarySpecCoordTransformGraph[RunType](
+        {
+            "a4": detector_geometric_a4,
+            "analyzer_dspacing": lambda: analyzer["dspacing"],
+            "analyzer_position": lambda: analyzer["position"],
+            "analyzer_transform": lambda: analyzer["transform"],
+            "beam_aligned_unit_vectors": beam_aligned_unit_vectors,
+            "detector_position": "position",
+            "sample_analyzer_vector": sample_analyzer_vector,
+            "analyzer_detector_vector": analyzer_detector_vector,
+            "final_energy": final_energy,
+            "final_wavenumber": final_wavenumber,
+            "final_wavevector": final_wavevector,
+            "L2": secondary_flight_path_length,
+            "secondary_flight_time": secondary_flight_time,
+        }
+    )
+
+
+def move_time_to_sample(
+    data: DataGroupedByRotation[RunType], pulse_period: PulsePeriod
+) -> DataAtSample[RunType]:
+    """Return the events with the event_time_offset coordinate offset to time at sample.
+
+    Parameters
+    ----------
+    data:
+        Data array with "event_time_offset" and "secondary_flight_time" coordinates.
+    pulse_period:
+        Duration of a neutron pulse.
+
+    Returns
+    -------
+    :
+        A shallow copy of ``data`` where the "event_time_offset" coordinate has been
+        shifted to the time at the sample.
+    """
+    offset = in_same_unit(
+        data.coords['secondary_flight_time'], data.bins.coords['event_time_offset']
+    )
+    time = data.bins.coords['event_time_offset'] - offset
+    time %= in_same_unit(pulse_period, time)
+    return DataAtSample[RunType](
+        data
+        # These are the detector positions and they no longer match the time:
+        .drop_coords(('position', 'x_pixel_offset', 'y_pixel_offset'))
```

Comment:
Is there no `z_pixel_offset`?

Reply:
No, there isn't. The file doesn't have any z offsets. I guess they are just 0.

Reply:
The pixel offsets are relative to the center of the triplet in a per-triplet coordinate system where the tubes are in the (x,y) plane -- so yes, there are no `z_pixel_offset` values specified because they're all 0.